File: Vue/src/utils/BgEffectDataManager.ts
export interface BgEffectData {
	uTranslateY: number;
	uPoints: number[];
	uAlphaMulti: number;
	uNoiseScale: number;
	uPointOffset: number;
	uPointRadiusMulti: number;
	uSaturateOffset: number;
	uLightOffset: number;
	uAlphaOffset: number;
	uShadowColorMulti: number;
	uShadowColorOffset: number;
	uShadowNoiseScale: number;
	uShadowOffset: number;
	colorInterpPeriod: number;
	gradientSpeedChange: number;
	gradientSpeedRest: number;
	gradientColors1: number[];
	gradientColors2: number[];
	gradientColors3: number[];
}

export type DeviceType = "PHONE" | "TABLET";
export const DeviceType = {
	PHONE: "PHONE" as DeviceType,
	TABLET: "TABLET" as DeviceType,
};

export type ThemeMode = "LIGHT" | "DARK";
export const ThemeMode = {
	LIGHT: "LIGHT" as ThemeMode,
	DARK: "DARK" as ThemeMode,
};

export class BgEffectDataManager {
	private dataPhoneLight: BgEffectData;
	private dataPadLight: BgEffectData;
	private dataPhoneDark: BgEffectData;
	private dataPadDark: BgEffectData;

	constructor() {
		this.dataPhoneLight = {
			uTranslateY: 0.0,
			uPoints: [0.8, 0.2, 1.0, 0.8, 0.9, 1.0, 0.2, 0.9, 1.0, 0.2, 0.2, 1.0],
			uAlphaMulti: 1.0,
			uNoiseScale: 1.5,
			uPointOffset: 0.2,
			uPointRadiusMulti: 1.0,
			uSaturateOffset: 0.2,
			uLightOffset: 0.1,
			uAlphaOffset: 0.5,
			uShadowColorMulti: 0.3,
			uShadowColorOffset: 0.3,
			uShadowNoiseScale: 5.0,
			uShadowOffset: 0.01,
			colorInterpPeriod: 5.0,
			gradientSpeedChange: 1.6,
			gradientSpeedRest: 1.05,
			gradientColors1: [1.0, 0.9, 0.94, 1.0, 1.0, 0.84, 0.89, 1.0, 0.97, 0.73, 0.82, 1.0, 0.64, 0.65, 0.98, 1.0],
			gradientColors2: [0.58, 0.74, 1.0, 1.0, 1.0, 0.9, 0.93, 1.0, 0.74, 0.76, 1.0, 1.0, 0.97, 0.77, 0.84, 1.0],
			gradientColors3: [0.98, 0.86, 0.9, 1.0, 0.6, 0.73, 0.98, 1.0, 0.92, 0.93, 1.0, 1.0, 0.56, 0.69, 1.0, 1.0],
		};

		this.dataPadLight = {
			uTranslateY: 0.0,
			uPoints: [0.8, 0.2, 1.0, 0.8, 0.9, 1.0, 0.2, 0.9, 1.0, 0.2, 0.2, 1.0],
			uAlphaMulti: 1.0,
			uNoiseScale: 1.5,
			uPointOffset: 0.2,
			uPointRadiusMulti: 1.0,
			uSaturateOffset: 0.2,
			uLightOffset: 0.1,
			uAlphaOffset: 0.5,
			uShadowColorMulti: 0.3,
			uShadowColorOffset: 0.3,
			uShadowNoiseScale: 5.0,
			uShadowOffset: 0.01,
			colorInterpPeriod: 7.0,
			gradientSpeedChange: 1.8,
			gradientSpeedRest: 1.0,
			gradientColors1: [0.99, 0.77, 0.86, 1.0, 0.74, 0.76, 1.0, 1.0, 0.72, 0.74, 1.0, 1.0, 0.98, 0.76, 0.8, 1.0],
			gradientColors2: [0.66, 0.75, 1.0, 1.0, 1.0, 0.86, 0.91, 1.0, 0.74, 0.76, 1.0, 1.0, 0.97, 0.77, 0.84, 1.0],
			gradientColors3: [0.97, 0.79, 0.85, 1.0, 0.65, 0.68, 0.98, 1.0, 0.66, 0.77, 1.0, 1.0, 0.72, 0.73, 0.98, 1.0],
		};

		this.dataPhoneDark = {
			uTranslateY: 0.0,
			uPoints: [0.8, 0.2, 1.0, 0.8, 0.9, 1.0, 0.2, 0.9, 1.0, 0.2, 0.2, 1.0],
			uAlphaMulti: 1.0,
			uNoiseScale: 1.5,
			uPointOffset: 0.4,
			uPointRadiusMulti: 1.0,
			uSaturateOffset: 0.17,
			uLightOffset: 0.0,
			uAlphaOffset: 0.5,
			uShadowColorMulti: 0.3,
			uShadowColorOffset: 0.3,
			uShadowNoiseScale: 5.0,
			uShadowOffset: 0.01,
			colorInterpPeriod: 8.0,
			gradientSpeedChange: 1.0,
			gradientSpeedRest: 1.0,
			gradientColors1: [0.2, 0.06, 0.88, 0.4, 0.3, 0.14, 0.55, 0.5, 0.0, 0.64, 0.96, 0.5, 0.11, 0.16, 0.83, 0.4],
			gradientColors2: [0.07, 0.15, 0.79, 0.5, 0.62, 0.21, 0.67, 0.5, 0.06, 0.25, 0.84, 0.5, 0.0, 0.2, 0.78, 0.5],
			gradientColors3: [0.58, 0.3, 0.74, 0.4, 0.27, 0.18, 0.6, 0.5, 0.66, 0.26, 0.62, 0.5, 0.12, 0.16, 0.7, 0.6],
		};

		this.dataPadDark = {
			uTranslateY: 0.0,
			uPoints: [0.8, 0.2, 1.0, 0.8, 0.9, 1.0, 0.2, 0.9, 1.0, 0.2, 0.2, 1.0],
			uAlphaMulti: 1.0,
			uNoiseScale: 1.5,
			uPointOffset: 0.2,
			uPointRadiusMulti: 1.0,
			uSaturateOffset: 0.0,
			uLightOffset: 0.0,
			uAlphaOffset: 0.5,
			uShadowColorMulti: 0.3,
			uShadowColorOffset: 0.3,
			uShadowNoiseScale: 5.0,
			uShadowOffset: 0.01,
			colorInterpPeriod: 7.0,
			gradientSpeedChange: 1.6,
			gradientSpeedRest: 1.2,
			gradientColors1: [0.66, 0.26, 0.62, 0.4, 0.06, 0.25, 0.84, 0.5, 0.0, 0.64, 0.96, 0.5, 0.14, 0.18, 0.55, 0.5],
			gradientColors2: [0.07, 0.15, 0.79, 0.5, 0.11, 0.16, 0.83, 0.5, 0.06, 0.25, 0.84, 0.5, 0.66, 0.26, 0.62, 0.5],
			gradientColors3: [0.58, 0.3, 0.74, 0.5, 0.11, 0.16, 0.83, 0.5, 0.66, 0.26, 0.62, 0.5, 0.27, 0.18, 0.6, 0.6],
		};
	}

	getData(deviceType: DeviceType, themeMode: ThemeMode): BgEffectData {
		if (deviceType === DeviceType.PHONE) {
			return themeMode === ThemeMode.LIGHT ? this.dataPhoneLight : this.dataPhoneDark;
		} else {
			return themeMode === ThemeMode.LIGHT ? this.dataPadLight : this.dataPadDark;
		}
	}
}

File: KMP_Js/composeApp/src/jsMain/kotlin/plus/yumeyuka/yumeyuka/BgEffectDataManager.kt
// https://github.com/ReChronoRain/HyperCeiler/tree/main/app/src/main/java/com/sevtinge/hyperceiler/main/page/about/controller

package plus.yumeyuka.yumeyuka

import androidx.compose.ui.graphics.Color

enum class DeviceType {
    PHONE,
    TABLET
}

enum class ThemeMode {
    LIGHT,
    DARK
}

data class BgEffectData(
    var uTranslateY: Float = 0.0f,
    var uPoints: FloatArray = floatArrayOf(),
    var uAlphaMulti: Float = 1.0f,
    var uNoiseScale: Float = 1.5f,
    var uPointOffset: Float = 0.2f,
    var uPointRadiusMulti: Float = 1.0f,
    var uSaturateOffset: Float = 0.2f,
    var uLightOffset: Float = 0.1f,
    var uAlphaOffset: Float = 0.5f,
    var uShadowColorMulti: Float = 0.3f,
    var uShadowColorOffset: Float = 0.3f,
    var uShadowNoiseScale: Float = 5.0f,
    var uShadowOffset: Float = 0.01f,
    var colorInterpPeriod: Float = 5.0f,
    var gradientSpeedChange: Float = 1.6f,
    var gradientSpeedRest: Float = 1.05f,
    var gradientColors1: Array<FloatArray> = emptyArray(),
    var gradientColors2: Array<FloatArray> = emptyArray(),
    var gradientColors3: Array<FloatArray> = emptyArray()
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || this::class != other::class) return false
        other as BgEffectData
        return uPoints.contentEquals(other.uPoints)
    }

    override fun hashCode(): Int {
        return uPoints.contentHashCode()
    }
}

class BgEffectDataManager {
    val dataPhoneLight: BgEffectData
    val dataPadLight: BgEffectData
    val dataPhoneDark: BgEffectData
    val dataPadDark: BgEffectData

    init {
        dataPhoneLight = BgEffectData().apply {
            uTranslateY = 0.0f
            uPoints = floatArrayOf(0.8f, 0.2f, 1.0f, 0.8f, 0.9f, 1.0f, 0.2f, 0.9f, 1.0f, 0.2f, 0.2f, 1.0f)
            uAlphaMulti = 1.0f
            uNoiseScale = 1.5f
            uPointOffset = 0.2f
            uPointRadiusMulti = 1.0f
            uSaturateOffset = 0.2f
            uLightOffset = 0.1f
            uAlphaOffset = 0.5f
            uShadowColorMulti = 0.3f
            uShadowColorOffset = 0.3f
            uShadowNoiseScale = 5.0f
            uShadowOffset = 0.01f
            colorInterpPeriod = 10.0f
            gradientSpeedChange = 1.6f
            gradientSpeedRest = 1.05f

            gradientColors1 = arrayOf(
                floatArrayOf(1.0f, 0.9f, 0.94f, 1.0f),
                floatArrayOf(1.0f, 0.84f, 0.89f, 1.0f),
                floatArrayOf(0.97f, 0.73f, 0.82f, 1.0f),
                floatArrayOf(0.64f, 0.65f, 0.98f, 1.0f)
            )

            gradientColors2 = arrayOf(
                floatArrayOf(0.58f, 0.74f, 1.0f, 1.0f),
                floatArrayOf(1.0f, 0.9f, 0.93f, 1.0f),
                floatArrayOf(0.74f, 0.76f, 1.0f, 1.0f),
                floatArrayOf(0.97f, 0.77f, 0.84f, 1.0f)
            )

            gradientColors3 = arrayOf(
                floatArrayOf(0.98f, 0.86f, 0.9f, 1.0f),
                floatArrayOf(0.6f, 0.73f, 0.98f, 1.0f),
                floatArrayOf(0.92f, 0.93f, 1.0f, 1.0f),
                floatArrayOf(0.56f, 0.69f, 1.0f, 1.0f)
            )
        }

        dataPadLight = BgEffectData().apply {
            uTranslateY = 0.0f
            uPoints = floatArrayOf(0.8f, 0.2f, 1.0f, 0.8f, 0.9f, 1.0f, 0.2f, 0.9f, 1.0f, 0.2f, 0.2f, 1.0f)
            uAlphaMulti = 1.0f
            uNoiseScale = 1.5f
            uPointOffset = 0.2f
            uPointRadiusMulti = 1.0f
            uSaturateOffset = 0.2f
            uLightOffset = 0.1f
            uAlphaOffset = 0.5f
            uShadowColorMulti = 0.3f
            uShadowColorOffset = 0.3f
            uShadowNoiseScale = 5.0f
            uShadowOffset = 0.01f
            colorInterpPeriod = 10.0f
            gradientSpeedChange = 1.8f
            gradientSpeedRest = 1.0f

            gradientColors1 = arrayOf(
                floatArrayOf(0.99f, 0.77f, 0.86f, 1.0f),
                floatArrayOf(0.74f, 0.76f, 1.0f, 1.0f),
                floatArrayOf(0.72f, 0.74f, 1.0f, 1.0f),
                floatArrayOf(0.98f, 0.76f, 0.8f, 1.0f)
            )

            gradientColors2 = arrayOf(
                floatArrayOf(0.66f, 0.75f, 1.0f, 1.0f),
                floatArrayOf(1.0f, 0.86f, 0.91f, 1.0f),
                floatArrayOf(0.74f, 0.76f, 1.0f, 1.0f),
                floatArrayOf(0.97f, 0.77f, 0.84f, 1.0f)
            )

            gradientColors3 = arrayOf(
                floatArrayOf(0.97f, 0.79f, 0.85f, 1.0f),
                floatArrayOf(0.65f, 0.68f, 0.98f, 1.0f),
                floatArrayOf(0.66f, 0.77f, 1.0f, 1.0f),
                floatArrayOf(0.72f, 0.73f, 0.98f, 1.0f)
            )
        }

        dataPhoneDark = BgEffectData().apply {
            uTranslateY = 0.0f
            uPoints = floatArrayOf(0.8f, 0.2f, 1.0f, 0.8f, 0.9f, 1.0f, 0.2f, 0.9f, 1.0f, 0.2f, 0.2f, 1.0f)
            uAlphaMulti = 1.0f
            uNoiseScale = 1.5f
            uPointOffset = 0.4f
            uPointRadiusMulti = 1.0f
            uSaturateOffset = 0.17f
            uLightOffset = 0.0f
            uAlphaOffset = 0.5f
            uShadowColorMulti = 0.3f
            uShadowColorOffset = 0.3f
            uShadowNoiseScale = 5.0f
            uShadowOffset = 0.01f
            colorInterpPeriod = 10.0f
            gradientSpeedChange = 1.0f
            gradientSpeedRest = 1.0f

            gradientColors1 = arrayOf(
                floatArrayOf(0.2f, 0.06f, 0.88f, 0.4f),
                floatArrayOf(0.3f, 0.14f, 0.55f, 0.5f),
                floatArrayOf(0.0f, 0.64f, 0.96f, 0.5f),
                floatArrayOf(0.11f, 0.16f, 0.83f, 0.4f)
            )

            gradientColors2 = arrayOf(
                floatArrayOf(0.07f, 0.15f, 0.79f, 0.5f),
                floatArrayOf(0.62f, 0.21f, 0.67f, 0.5f),
                floatArrayOf(0.06f, 0.25f, 0.84f, 0.5f),
                floatArrayOf(0.0f, 0.2f, 0.78f, 0.5f)
            )

            gradientColors3 = arrayOf(
                floatArrayOf(0.58f, 0.3f, 0.74f, 0.4f),
                floatArrayOf(0.27f, 0.18f, 0.6f, 0.5f),
                floatArrayOf(0.66f, 0.26f, 0.62f, 0.5f),
                floatArrayOf(0.12f, 0.16f, 0.7f, 0.6f)
            )
        }

        dataPadDark = BgEffectData().apply {
            uTranslateY = 0.0f
            uPoints = floatArrayOf(0.8f, 0.2f, 1.0f, 0.8f, 0.9f, 1.0f, 0.2f, 0.9f, 1.0f, 0.2f, 0.2f, 1.0f)
            uAlphaMulti = 1.0f
            uNoiseScale = 1.5f
            uPointOffset = 0.2f
            uPointRadiusMulti = 1.0f
            uSaturateOffset = 0.0f
            uLightOffset = 0.0f
            uAlphaOffset = 0.5f
            uShadowColorMulti = 0.3f
            uShadowColorOffset = 0.3f
            uShadowNoiseScale = 5.0f
            uShadowOffset = 0.01f
            colorInterpPeriod = 10.0f
            gradientSpeedChange = 1.6f
            gradientSpeedRest = 1.2f

            gradientColors1 = arrayOf(
                floatArrayOf(0.66f, 0.26f, 0.62f, 0.4f),
                floatArrayOf(0.06f, 0.25f, 0.84f, 0.5f),
                floatArrayOf(0.0f, 0.64f, 0.96f, 0.5f),
                floatArrayOf(0.14f, 0.18f, 0.55f, 0.5f)
            )

            gradientColors2 = arrayOf(
                floatArrayOf(0.07f, 0.15f, 0.79f, 0.5f),
                floatArrayOf(0.11f, 0.16f, 0.83f, 0.5f),
                floatArrayOf(0.06f, 0.25f, 0.84f, 0.5f),
                floatArrayOf(0.66f, 0.26f, 0.62f, 0.5f)
            )

            gradientColors3 = arrayOf(
                floatArrayOf(0.58f, 0.3f, 0.74f, 0.5f),
                floatArrayOf(0.11f, 0.16f, 0.83f, 0.5f),
                floatArrayOf(0.66f, 0.26f, 0.62f, 0.5f),
                floatArrayOf(0.27f, 0.18f, 0.6f, 0.6f)
            )
        }
    }

    fun getData(deviceType: DeviceType, themeMode: ThemeMode): BgEffectData {
        return when (deviceType) {
            DeviceType.PHONE -> if (themeMode == ThemeMode.LIGHT) dataPhoneLight else dataPhoneDark
            DeviceType.TABLET -> if (themeMode == ThemeMode.LIGHT) dataPadLight else dataPadDark
        }
    }
}

fun FloatArray.toComposeColor(): Color {
    return Color(this[0], this[1], this[2], this[3])
}